allennlp/training/metrics/srl_eval_scorer.py

```diff
@@ -0,0 +1,159 @@
+        self._false_positives: Dict[str, int] = defaultdict(int)
+        self._false_negatives: Dict[str, int] = defaultdict(int)
+
+    @overrides
+    def __call__(self,  # type: ignore
+                 batch_verb_indices: List[Optional[int]],
+                 batch_sentences: List[List[str]],
+                 batch_conll_formatted_predicted_tags: List[List[str]],
+                 batch_conll_formatted_gold_tags: List[List[str]]) -> None:
+        # pylint: disable=signature-differs
+        """
+        Parameters
+        ----------
+        batch_verb_indices : ``List[Optional[int]]``, required.
+            The indices of the verbal predicate in the sentences which
+            the gold labels are the arguments for, or None if the sentence
+            contains no verbal predicate.
+        batch_sentences : ``List[List[str]]``, required.
+            The word tokens for each instance in the batch.
+        batch_conll_formatted_predicted_tags : ``List[str]``, required.
+            A list of predicted CoNLL-formatted SRL tags (itself a list) to compute score for.
+            Use allennlp.models.semantic_role_labeler.convert_bio_tags_to_conll_format
+            to convert from BIO to CoNLL format before passing the tags into the metric,
+            if applicable.
+        batch_conll_formatted_gold_tags : ``List[str]``, required.
+            A list of gold CoNLL-formatted SRL tags (itself a list) to use as a reference.
+            Use allennlp.models.semantic_role_labeler.convert_bio_tags_to_conll_format
+            to convert from BIO to CoNLL format before passing the
+            tags into the metric, if applicable.
+        """
+        if not os.path.exists(self._srl_eval_path):
+            raise ConfigurationError(f"srl-eval.pl not found at {self._srl_eval_path}.")
+        tempdir = tempfile.mkdtemp()
+        gold_path = os.path.join(tempdir, "gold.txt")
+        predicted_path = os.path.join(tempdir, "predicted.txt")
+        output_path = os.path.join(tempdir, "output.txt")
+
+        with open(predicted_path, "w") as predicted_file, open(gold_path, "w") as gold_file:
+            for verb_index, sentence, predicted_tag_sequence, gold_tag_sequence in zip(
+                    batch_verb_indices,
+                    batch_sentences,
+                    batch_conll_formatted_predicted_tags,
+                    batch_conll_formatted_gold_tags):
+                write_conll_formatted_tags_to_file(predicted_file,
+                                                   gold_file,
+                                                   verb_index,
+                                                   sentence,
+                                                   predicted_tag_sequence,
+                                                   gold_tag_sequence)
+        perl_script_command = (f"perl {self._srl_eval_path} {predicted_path} {gold_path} "
+                               f"> {output_path}")
```

Comment:
Rather than redirecting to a file you can directly capture the stdout. Should be something like:

```
completed_process = subprocess.run(["perl", self._srl_eval_path, predicted_path, gold_path], check=True, text=True)
for line in completed_process.stdout.split("\n"):
...
```

(Untested, so do check the docs: https://docs.python.org/3/library/subprocess.html#subprocess.CompletedProcess)

This may help address your comment above about writing to disk often. In general it's also nice to avoid shell expansion for security's sake. (Though I think all the paths here should be controlled by the developer, so they should be safe...)

Reply:
ok, good to know! We should probably also fix this in the evalb metric, then, but that's a separate issue: https://github.com/allenai/allennlp/blob/master/allennlp/training/metrics/evalb_bracketing_scorer.py#L96-L98